### PR TITLE
Implemented Cullum/Willoughby method for removing spurious eigenvalues.

### PR DESCRIPTION
**Pull request by aprokop**
This is a cleanup of #143. It is to be served as to improve status quo while continuing investigation of potential issues in that PR.

### CODE REVIEW

File: include/mfmg/dealii/lanczos.templates.hpp [hunk 1]
```diff
@@ -231,6 +231,8 @@ Lanczos<OperatorType, VectorType>::details_solve_lanczos(
       it_prev_check = it;
     }
   }
+  it = it < maxit ? it : maxit;
```
Comment (masterleinad) [inline include/mfmg/dealii/lanczos.templates.hpp]:
```suggestion
  it = std::min(it, maxit);
```

**Comment by masterleinad** [comment]
So can we close #143?

**Comment by aprokop** [comment]
> So can we close #143?

No, we have not solved all the issues, and Wayne will continue using it for the development. This PR is simply cleaning up that and merging so that we have something better than what we have right now.

**Comment by aprokop** [comment]
What's with the @ghost account assigning labels? Something happened to waffle?

**Comment by Rombur** [comment]
Yeah Waffle is dead. Got an email a month ago saying that they would close.

File: include/mfmg/dealii/lanczos.templates.hpp [hunk 2]
```diff
@@ -284,6 +288,104 @@ Lanczos<OperatorType, VectorType>::details_calc_tridiag_epairs(
...
+  const double tol2 = 5.e-12;
+
+  // Identify repeated eigenvalues.
+  // A "marked" eigenvalue here is nonrepeated or first of a set of repeated.
+  std::vector<bool> is_repeated(n);
+  std::vector<bool> is_marked(n);
+  for (int i = 0; i < n; ++i)
+  {
+    is_repeated[i] = ((i > 0 && (evals[i] <= evals[i - 1] + tol2)) ||
+                      (i < n - 1 && (evals[i + 1] <= evals[i] + tol2)));
```
Comment (Rombur) [inline include/mfmg/dealii/lanczos.templates.hpp]: Can you add a comment on this condition. This is really hard to read.

File: include/mfmg/dealii/lanczos.templates.hpp [hunk 3]
```diff
@@ -284,6 +288,104 @@ Lanczos<OperatorType, VectorType>::details_calc_tridiag_epairs(
...
+    is_marked[i] = (i == 0 || (evals[i] > evals[i - 1] + tol2));
```
Comment (Rombur) [inline include/mfmg/dealii/lanczos.templates.hpp]: Same here.

**Comment by aprokop** [comment]
Rebased and added comments.

**Pull request merged**